panic: runtime error: slice bounds out of range in `(*Hive).GetValueData`, resident data length not validated

**Package:** `github.com/mandiant/gopacket/pkg/registry`  
**Affected file:** `pkg/registry/hive.go`  
**Commit tested:** `5d927b8`

## Summary

`(*Hive).GetValueData` can panic with a slice-bounds-out-of-range error when parsing a registry hive whose `VKRecord.DataLen` encodes a resident value with a length greater than 4. Because `DataLen` is read verbatim from hive bytes with no prior range check, any caller that parses an attacker-supplied hive file is exposed to a denial-of-service crash.

## Root cause

The Windows Registry format uses bit 31 of `VKRecord.DataLen` as a "resident" flag: when set, the actual data (up to 4 bytes) is stored inline in `DataOffset` rather than in a separate cell. The relevant code in `GetValueData` is:

```go
// Data is stored in the DataOffset field itself (up to 4 bytes)
data := make([]byte, 4)
binary.LittleEndian.PutUint32(data, vk.DataOffset)
return data[:dataLen], nil
```

The comment documents the invariant ("up to 4 bytes"), but no guard enforces it. `dataLen` is derived as `vk.DataLen & 0x7FFFFFFF`, and `vk.DataLen` is populated by `parseVK` directly from raw hive bytes:

```go
binary.Read(r, binary.LittleEndian, &vk.DataLen)  // no range check
```

If the hive contains `DataLen = 0x80000005` (resident flag set, lower 31 bits = 5), then `dataLen` becomes 5 and `data[:5]` panics on the 4-byte slice.

## Steps to reproduce

```go
package main

import (
    "fmt"
    "encoding/binary"
    "github.com/mandiant/gopacket/pkg/registry"
)

func main() {
    // Minimal valid hive: regf magic + rootOffset at offset 36
    hive := make([]byte, 4096+64)
    copy(hive, []byte{0x72, 0x65, 0x67, 0x66}) // "regf"

    h, err := registry.Open(hive)
    if err != nil {
        panic(err)
    }

    vk := &registry.VKRecord{
        Signature:  0x6B76,
        DataLen:    0x80000005, // isResident=true, dataLen=5 --> data[:5] on a 4-byte slice
        DataOffset: 0x01020304,
        DataType:   1,
    }

    data, err := h.GetValueData(vk) // panic: runtime error: slice bounds out of range [:5] with capacity 4
    fmt.Println(data, err)
}
```

**Panic output:**

```
panic: runtime error: slice bounds out of range [:5] with capacity 4

goroutine 1 [running]:
github.com/mandiant/gopacket/pkg/registry.(*Hive).GetValueData(0x37f9430a0000?, 0x37f943092f10)
	/tmp/gopath1113421067/pkg/mod/github.com/mandiant/gopacket@v0.0.0-20260424163850-5d927b8e6b8d/pkg/registry/hive.go:258 +0xb0
main.main()
	/tmp/sandbox3190411761/prog.go:26 +0x9f
```

## How the bug was found

This bug was discovered using [Zorya](https://github.com/0xchase/zorya), a concolic execution engine for ELF binaries. Zorya combines a GDB-based concrete execution snapshot with P-code symbolic execution and the Z3 SMT solver to find inputs that drive a program to a crash site.

A test harness was written that called the resident-data path of `GetValueData` directly, passing `dataLen` and `dataOffset` as arguments:

```go
//go:noinline
func exerciseResident(dataLen uint32, dataOffset uint32) {
    data := make([]byte, 4)
    data[0] = byte(dataOffset)
    // ...
    result := data[:dataLen] // target: slice bounds panic
    fmt.Printf("resident ok len=%d\n", len(result))
}
```
Zorya was pointed at this function in `--mode function`, which snapshotted the process at the function entry, promoted `dataLen` and `dataOffset` to symbolic Z3 variables, and ran concolic analysis:

```sh
zorya path/to/bin \
  --mode function 0x4a9560 \
  --lang go --compiler gc \
  --thread-scheduling main-only \
  --arg "72656766000000000000000000000000000000000000000000000000000000000000000000000000" \
  --negate-path-exploration
```

(`0x4a9560` is the address of `main.exerciseResident` in the compiled harness binary; `--arg` is a minimal valid regf header that lets the harness reach the function without error.)

Within 44 seconds it produced the following witness:

```
[*] SATISFIABLE STATE FOUND
Timestamp: 2026-05-11 08:02:00 UTC
Elapsed since start: 44.661s
Instruction Address: 0x4a9574
Panic Address:       0x4a95cb
Opcode:              CBRANCH
Detection method:    Exploring the not taken path with Overlay Execution

The program can panic if its inputs are the following:
  - The input 'dataLen' must be 5 (unsigned: 5; signed: 5)
```

The CBRANCH oracle detected that the slice bounds check at `0x4a9574` branches to a `runtime.panicSliceB` call site (`0x4a95cb`) when `dataLen > 4`, and Z3 solved the single constraint `dataLen ≥ 5` to produce the minimal witness `dataLen = 5`.

## Impact

Any application that uses this package to parse registry hives from untrusted sources (forensics pipelines, EDR agents, malware analysis tooling) can be crashed by providing a single malformed VK record with `DataLen = 0x80000005`. The crash is deterministic and requires no special privileges.

## Suggested fix

Add a bounds check before the slice expression:

```go
if isResident {
    if dataLen > 4 {
        return nil, fmt.Errorf("resident data length %d exceeds maximum of 4 bytes", dataLen)
    }
    data := make([]byte, 4)
    binary.LittleEndian.PutUint32(data, vk.DataOffset)
    return data[:dataLen], nil
}
```

This is consistent with the existing comment ("up to 4 bytes") and matches the Windows Registry format specification, which states that inline resident values are limited to 4 bytes.

--- a/pkg/registry/hive.go
+++ b/pkg/registry/hive.go
@@ -252,7 +252,13 @@ func (h *Hive) GetValueData(vk *VKRecord) ([]byte, error) {
 	isResident := vk.DataLen&0x80000000 != 0
 
 	if isResident {
-		// Data is stored in the DataOffset field itself (up to 4 bytes)
+		// Data is stored in the DataOffset field itself (up to 4 bytes).
+		// dataLen is attacker-controlled (lower 31 bits of vk.DataLen, read
+		// verbatim from hive bytes); reject lengths beyond the 4-byte cap
+		// before slicing to prevent a panic on malformed hives.
+		if dataLen > 4 {
+			return nil, fmt.Errorf("resident data length %d exceeds maximum of 4 bytes", dataLen)
+		}
 		data := make([]byte, 4)
 		binary.LittleEndian.PutUint32(data, vk.DataOffset)
 		return data[:dataLen], nil
@@ -376,9 +382,15 @@ func (h *Hive) enumSubKeys(nk *NKRecord) ([]subKeyInfo, error) {
 			}
 			subListOffset := int32(binary.LittleEndian.Uint32(cell[entryOff : entryOff+4]))
 
-			// Recursively process sub-list
+			// Recursively process sub-list. readCell can return a slice
+			// shorter than the 4-byte cell header (e.g. minimum-size cell with
+			// no usable bytes), so guard against an indexing panic before
+			// reading the sub-list signature and count.
 			subCell, err := h.readCell(subListOffset)
 			if err != nil {
+				continue
+			}
+			if len(subCell) < 4 {
 				continue
 			}
 
@@ -636,16 +648,32 @@ func (h *Hive) SetValueData(keyOffset int32, valueName string, newData []byte) e
 		if isResident {
 			// Data stored inline in the VK record's DataOffset field
 			// VK cell starts at cellOffset(vkOffset)+4 (skip cell size)
-			// DataOffset is at bytes 8..12 of the VK record (after sig:2 + nameLen:2 + dataLen:4)
+			// DataOffset is at bytes 8..12 of the VK record (after sig:2 + nameLen:2 + dataLen:4).
+			// Same 4-byte cap as the read path in GetValueData; without this
+			// guard a length-matching newData buffer of 5+ bytes would silently
+			// scribble into the adjacent hive cell.
+			if dataLen > 4 {
+				return fmt.Errorf("resident data length %d exceeds maximum of 4 bytes", dataLen)
+			}
 			vkPos := h.cellOffset(vkOffset) + 4 // skip cell size dword
 			dataFieldPos := vkPos + 8           // offset of DataOffset field
 			copy(h.data[dataFieldPos:dataFieldPos+int(dataLen)], newData)
 			return nil
 		}
 
-		// Data is in a separate cell; overwrite in place
-		dataPos := h.cellOffset(int32(vk.DataOffset)) + 4 // skip cell size
-		copy(h.data[dataPos:dataPos+int(dataLen)], newData)
+		// Data is in a separate cell; overwrite in place. vk.DataOffset is
+		// attacker-controlled in malformed hives, so route through readCell
+		// (which validates the cell header and bounds) and verify dataLen
+		// fits before mutating. Without this guard a hostile offset can
+		// scribble over arbitrary hive bytes including the regf header.
+		cell, err := h.readCell(int32(vk.DataOffset))
+		if err != nil {
+			return err
+		}
+		if int(dataLen) > len(cell) {
+			return fmt.Errorf("value data length %d exceeds cell size %d", dataLen, len(cell))
+		}
+		copy(cell[:dataLen], newData)
 		return nil
 	}
 

--- a/pkg/registry/hive_test.go
+++ b/pkg/registry/hive_test.go
@@ -86,3 +86,78 @@ func TestReadCellReturnsDataForValidCell(t *testing.T) {
 		t.Fatalf("readCell returned %q, want %q", string(data), "hello, world")
 	}
 }
+
+// TestGetValueDataResidentLengthRejected covers the input that originally
+// panicked in GetValueData: a VK with the resident flag set and a lower-31
+// length greater than the 4-byte DataOffset field. Found by kajaaz via Zorya
+// (issue #25). The guard must return an error, not crash.
+func TestGetValueDataResidentLengthRejected(t *testing.T) {
+	// h.data is unused on the resident path; the guard short-circuits before
+	// any cell read happens.
+	h := &Hive{}
+
+	for _, dataLen := range []uint32{5, 8, 0xFF, 0x7FFFFFFF} {
+		t.Run("dataLen="+strings.ToUpper(uintHex(dataLen)), func(t *testing.T) {
+			vk := &VKRecord{
+				DataLen:    dataLen | 0x80000000, // set resident flag
+				DataOffset: 0xDEADBEEF,
+			}
+
+			defer func() {
+				if r := recover(); r != nil {
+					t.Fatalf("GetValueData panicked: %v", r)
+				}
+			}()
+			data, err := h.GetValueData(vk)
+			if err == nil {
+				t.Fatalf("GetValueData returned nil error; got data=%v", data)
+			}
+			if !strings.Contains(err.Error(), "exceeds maximum of 4 bytes") {
+				t.Fatalf("GetValueData error %q did not contain expected substring", err.Error())
+			}
+		})
+	}
+}
+
+// TestGetValueDataResidentLengthAllowed confirms the happy path still works:
+// a resident value of 1..4 bytes returns the inline DataOffset bytes.
+func TestGetValueDataResidentLengthAllowed(t *testing.T) {
+	h := &Hive{}
+	// DataOffset 0x44434241 = bytes 'A','B','C','D' little-endian.
+	for dataLen, want := range map[uint32]string{
+		1: "A",
+		2: "AB",
+		3: "ABC",
+		4: "ABCD",
+	} {
+		t.Run("dataLen="+uintHex(dataLen), func(t *testing.T) {
+			vk := &VKRecord{
+				DataLen:    dataLen | 0x80000000,
+				DataOffset: 0x44434241,
+			}
+			data, err := h.GetValueData(vk)
+			if err != nil {
+				t.Fatalf("GetValueData error: %v", err)
+			}
+			if string(data) != want {
+				t.Fatalf("GetValueData returned %q, want %q", string(data), want)
+			}
+		})
+	}
+}
+
+// uintHex is a tiny helper for table-driven test names.
+func uintHex(v uint32) string {
+	const hex = "0123456789abcdef"
+	if v == 0 {
+		return "0"
+	}
+	var buf [10]byte
+	i := len(buf)
+	for v > 0 {
+		i--
+		buf[i] = hex[v&0xF]
+		v >>= 4
+	}
+	return string(buf[i:])
+}

